libflux/flux-core/src/semantic/bootstrap.rs

```diff
@@ -35,32 +36,89 @@ pub struct Error {
     pub msg: String,
 }
 
-/// DocPackage represents the documentation for a package and its sub packages
+/// Doc is an enum that can take the form of the various types of flux documentation structures through polymorphism.
 #[derive(Debug, Serialize, Deserialize)]
-pub struct DocPackage {
-    /// the path to the documentation package
-    pub path: String,
-    /// the name of the comments package?
+pub enum Doc {
+    /// Package represents documentation for an entire Flux package.
+    Package(Box<PackageDoc>),
+    /// Value represents documentation for a value exposed from a package.
+    Value(Box<ValueDoc>),
+    /// Builtin represents documentation for a builtin value exposed from a package.
+    Builtin(Box<ValueDoc>),
+    /// Option represents documentation for a option value exposed from a package.
+    Option(Box<ValueDoc>),
+    /// Function represents documentation for a function value exposed from a package.
+    Function(Box<FunctionDoc>),
+}
+
+/// PackageDoc represents the documentation for a package and its sub packages
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PackageDoc {
+    /// the name of the comments package
     pub name: String,
-    /// the doc
-    pub doc: String,
-    /// the values of the comments
-    pub values: Vec<DocValue>,
+    /// the headline of the package
+    pub headline: String,
+    /// the description of the package
+    pub description: Option<String>,
+    /// the members are the values and funcitons of a package
+    pub members: HashMap<String, Doc>,
```

Comment:
By using a hash map here, it implies that the order of the Doc items as they were in the Flux source won't be maintained. Maybe a `Vec` would be more appropriate here? I do not think any package contains so many items that performance is a concern.

Reply:
I think using a hashmap makes it less prone to errors when looping and also the faster lookup times is a plus. Its also more convenient to lookup a Doc by its name rather than by its index.

Reply:
I agree with @BarWeiner1's explanation here. As far as I know, we don't need to keep the docs sorted in any particular order, but I could be wrong. Is there a use case here that I'm not considering @wolffcm?

Reply:
Say for a site that is generated from the doc comments, we might want to preserve the order of the documentation that was in the Flux source? But we may also want to use alphabetical order or something like that. But it seems clear that we would not want the random order generated by iterating over a hash map.

If we want the items on a generated site to be in alphabetical order, then a hash map is fine, the site generator can take care of the sorting. But if we want to preserve the order of the Flux source, then we would lose the ability to do that. Maybe that's fine though, if this is not a requirement, feel free to disregard.

(I'm not saying that we couldn't also keep a hash map on the side to do fast lookups by name, probably you would want to support both.)

Reply:
That's fair. Maybe it's worth reaching out to the folks who are working on the new docs site. I can take care of that.

Reply:
Got some feedback. No interest in preserving the order found in the source code (that's not to say other consumers wouldn't want that at some point in the future), but there was some interest in grouping by type (i.e. All the functions in a package are grouped together, likewise for builtins, options, etc). I think that's a good idea, but it would mean making changes to the spec, and may be better suited for a separate issue. What do you think @wolffcm?

Reply:
Sounds good to me, thanks for following up.